### PR TITLE
Add documentation regarding authorization and access tokens

### LINKED ISSUE #10: Define authentication and access token architecture
- Codegarten API access - Design how a user authenticates via the client app and obtains an access token to use the CodeGarten API.

- GitHub API access - Design how the CodeGarten back-end obtains access tokens representing authorisations delegated by the user.

- What is the difference between a GitHub app and an OAuth app.

### PR DESCRIPTION
**Pull request by PTKay**
This PR adds information about the following things:
- How to obtain a user access token from GitHub
- Reasons as to why we chose to use a GitHub App instead of OAuth App
- How to manage GitHub App installations and how to obtain their IDs
- How to access the GitHub API
- How to manage CodeGarten API Access, including how to generate Access Tokens

Closes gh-10.

### CODE REVIEW

**Review by TiagoMMDavid — APPROVED** [review]
Looks great to me!

**Review by dvsshadow — APPROVED** [review]

File: docs/auth/app-comparison.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# Differences between GitHub App and OAuth App
+
+## Comparison
+* An OAuth application acts as an authenticated GitHub user (owner of the access token), while GitHub apps use their own identity
```
Comment (pmhsfelix) [inline docs/auth/app-comparison.md]: Suggestion: replace "act as" with "acts on behalf of"

File: docs/auth/app-comparison.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+* GitHub apps are installed in organizations (by an admin) or in an user, while OAuth apps are authorized to access user resources (e.g repositories that an user can access)
```
Comment (pmhsfelix) [inline docs/auth/app-comparison.md]: - "or in an user" - something is not right
- "OAuth apps are authorised **by users** t access resources as the user"

File: docs/auth/codegarten-api.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+In order to obtain a CodeGarten API Access Token, it's required to authenticate via GitHub. This process is described in the [corresponding documentation](github-api.md#user-access-token). It's also required to specify which client is requesting an access token by sending the client ID through the query string.
+
+After the CodeGarten IM gets a GitHub OAuth Access Token from the user, and after that token is stored in the database, the CodeGarten IM will generate a valid code to be exchanged by an API Access Token. This is possible due to the IM being contained inside the API. After generating the code, the IM will redirect the browser to the redirect URI specified in the query string of the authorization request. This redirect URI should be valid and trusted by the API. Finally, the App will then request an access token from the API, giving the code obtained through the CodeGarten IM and both the client ID and secret. If this is a valid code, and if the client information is correct, the API will then respond with an API Access Token, as well as giving its expiration date. The API will also store this access token in the database, associating it with an user and a client.
```
Comment (pmhsfelix) [inline docs/auth/codegarten-api.md]: What does "IM" mean?

File: docs/auth/github-api.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+### Installation Access Token
+To obtain an installation access token, the API will have an endpoint which will respond with the GitHub App installation URI, which will be used to install the GitHub App to an organization. The browser will then present the installation form to the user and, after the user picks and installs the app into an organization, the browser will be redirected to a new URI, where the installation ID is present in its query string. This endpoint is part of what we call **CodeGarten Interaction Manager (IM)**, which is neither the API nor a client app. Its only responsibility is to respond to requests such as GitHub Installations and GitHub/CodeGarten Authentications. Via this endpoint, the API will be able to receive the installation ID, and store this information in the database alongside the organization's ID. This interaction should be made on a new browser window, which will close itself in the end.
```
Comment (pmhsfelix) [inline docs/auth/github-api.md]: I don't understand the "which will respond with the GitHub App installation URI".

File: docs/auth/github-api.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+### User Access Token
+To obtain a user's access token, the API will have an endpoint which will respond with an URI from the CodeGarten IM module. This URI will redirect the browser to the GitHub App's OAuth URI, which in turn will request the user to authenticate itself in GitHub. After completing the authentication, the browser will be redirected back to CodeGarten IM, passing along a authorization code in the query string. This code will be exchanged for a GitHub OAuth Access Token related to the GitHub user. This access token will then be stored in the database alongside the user's information, replacing the old one if it exists.
```
Comment (pmhsfelix) [inline docs/auth/github-api.md:27]: An interaction diagram, preferably in text format, would be useful. Example:
```
- Client opens browser with an URI containing an authorization request to CG (CodeGarten)
- B --> CG/AS: GET authorization request-1, with scope=TODO
- B <-- CG/AS: redirect with an URI containing an authorization request to GH (GitHub), with scope=TODO
- B --> GH/AS: GET authorization-request-2
- B <-> GH/AS: interactive user authentication and consent
- B <-- GH/AS: redirect with Location=authorization-response-2
- B --> CG/AS: GET authorisation-response-2 (contains code issued by GH)
...
```
Also, this diagram should include what is done by internally by CG on each step.

**Review by pmhsfelix — CHANGES_REQUESTED** [review]
See comments. We need interaction/sequence diagram before proceeding.

File: docs/auth/codegarten-api.md [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    * Client ID - Used to identify what client is requesting an authorization code/access token
+    * Client Name - User friendly way of identifying the client
+    * Client Secret - Used when exchanging an authorization code for an access token
+    * Client Redirect URI - The URI that will receive the authorization code
+
+* CodeGarten API Access Token - Used to access the API endpoints that require authentication. It's associated with a user and a client
+* CodeGarten API Access Token Expiration Date - Used to determine when to request a new Access Token
+
+## Obtaining an Access Token
+In order to obtain a CodeGarten API Access Token, it's required to authenticate via GitHub. This process is described in the [corresponding documentation](github-api.md#user-access-token). It's also required to specify which client is requesting an access token by sending the client ID through the query string.
```
Comment (pmhsfelix) [inline docs/auth/codegarten-api.md:16]: "sending the client ID through the query string" - this follows the OAuth protocol, right?

File: docs/auth/codegarten-api.md [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+## Required Information
+In order to manage access to the CodeGarten API, it's required to store the following information:
+* CodeGarten Client - A client that's registered to use the API. This client will have:
+    * Client ID - Used to identify what client is requesting an authorization code/access token
+    * Client Name - User friendly way of identifying the client
+    * Client Secret - Used when exchanging an authorization code for an access token
+    * Client Redirect URI - The URI that will receive the authorization code
+
+* CodeGarten API Access Token - Used to access the API endpoints that require authentication. It's associated with a user and a client
+* CodeGarten API Access Token Expiration Date - Used to determine when to request a new Access Token
```
Comment (pmhsfelix) [inline docs/auth/codegarten-api.md:11]: - How is the client secret stored?
- How is the access token stored?

File: docs/auth/codegarten-api.md [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+After the **CodeGarten Interaction Manager (IM)** gets a GitHub OAuth Access Token from the user, and after that token is stored in the database, the CodeGarten IM will generate a valid code to be exchanged by an API Access Token. This is possible due to the IM being contained inside the API. After generating the code, the IM will redirect the browser to the redirect URI specified in the query string of the authorization request. This redirect URI should be valid and trusted by the API. Finally, the App will then request an access token from the API, giving the code obtained through the CodeGarten IM and both the client ID and secret. If this is a valid code, and if the client information is correct, the API will then respond with an API Access Token, as well as giving its expiration date. The API will also store this access token in the database, associating it with an user and a client.
```
Comment (pmhsfelix) [inline docs/auth/codegarten-api.md]: "gets a GitHub OAuth Access Token from the user". This means getting an access token from GitHub, based on an user delegation, right?

Comment (pmhsfelix) [reply docs/auth/codegarten-api.md]: How is the code stored and its use controlled?

**Review by pmhsfelix — APPROVED** [review]
Left some notes with additional information that should be present in the design document. Doesn't need to be done on this PR. Can be done in parallel with the implementation.

**Pull request merged**